File: apps/athenaeum/packages/domain/src/tag.ts
import * as Schema from "effect/Schema"
import { EntityId } from "./node.js"

// Plan §"Storage & domain model": "tags — supertags, multi-parent DAG: {id, name, parentIds[],
// builtin}. Fixed Base Tags (Person, Organization, Company, Event, Place, Area, Project, Task)
// seeded once at workspace creation as immutable `builtin: true` rows — mirrors GraphDataModel.md's
// 'Fixed Base Tags... ship with the application and are not uploaded.'"

export class Tag extends Schema.Class<Tag>("Tag")({
  id: EntityId,
  name: Schema.String.pipe(Schema.minLength(1)),
  parentIds: Schema.Array(EntityId),
  builtin: Schema.Boolean
}) {}

// --- Base Tags ---------------------------------------------------------------------------
//
// Tag-ID scheme decision (plan task item 2 — "decide and document whether IDs are literal
// well-known constants like 'tag-person' or generated, and be consistent"):
//
// `Tag.id` is typed as `EntityId`, which (per node.ts) only accepts a ULID or a UUID — a
// literal slug like `"tag-person"` fails that schema's own validation, so "literal slug
// constants" isn't actually available without weakening `EntityId` for every other entity in
// this package. Fully-random UUIDs would be deterministic-*able* (a UUIDv5 derived from a fixed
// namespace + slug) but that requires a hashing/UUID-library dependency this zero-dependency
// package (see package.json: only `effect`) doesn't have, just to compute eight fixed strings
// once.
//
// The chosen scheme: literal **nil-pattern UUIDs**, `00000000-0000-0000-0000-00000000000N` for
// N = 1..8 in the plan's own listed order (Person, Organization, Company, Event, Place, Area,
// Project, Task). This is deterministic (no computation, no library, identical across every
// workspace by construction), valid against the existing `EntityId` UUID pattern with no schema
// changes, and immediately recognizable as a reserved/builtin value in logs, DB dumps, and test
// fixtures — a real random-looking UUID could not be distinguished from an ordinary row's ID at
// a glance, which matters for a "ship with the application, not uploaded" fixed set. The backend
// seeds these same literal IDs at workspace-creation time (plan: "seeded once at workspace creation");
// nothing is assigned or generated at runtime.
export const BaseTagIds = {
  Person: EntityId.make("00000000-0000-0000-0000-000000000001"),
  Organization: EntityId.make("00000000-0000-0000-0000-000000000002"),
  Company: EntityId.make("00000000-0000-0000-0000-000000000003"),
  Event: EntityId.make("00000000-0000-0000-0000-000000000004"),
  Place: EntityId.make("00000000-0000-0000-0000-000000000005"),
  Area: EntityId.make("00000000-0000-0000-0000-000000000006"),
  Project: EntityId.make("00000000-0000-0000-0000-000000000007"),
  Task: EntityId.make("00000000-0000-0000-0000-000000000008")
} as const

/**
 * The 8 Base Tags, seeded once at workspace creation as immutable `builtin: true` rows (plan
 * §"Storage & domain model"). None have parents — the Base Tags are the roots of the supertag
 * DAG; user-created tags may later declare one or more of these (or each other) as `parentIds`.
 */
export const BASE_TAGS: ReadonlyArray<Tag> = [
  new Tag({ id: BaseTagIds.Person, name: "Person", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Organization, name: "Organization", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Company, name: "Company", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Event, name: "Event", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Place, name: "Place", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Area, name: "Area", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Project, name: "Project", parentIds: [], builtin: true }),
  new Tag({ id: BaseTagIds.Task, name: "Task", parentIds: [], builtin: true })
]

File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/Tag.swift
import Foundation

/// Mirrors `packages/domain/src/tag.ts`'s `Tag` — supertags, multi-parent DAG:
/// `{id, name, parentIds[], builtin}`.
public struct Tag: Codable, Hashable, Sendable {
    public let id: EntityId
    public let name: String
    public let parentIds: [EntityId]
    public let builtin: Bool

    public init(id: EntityId, name: String, parentIds: [EntityId], builtin: Bool) {
        self.id = id
        self.name = name
        self.parentIds = parentIds
        self.builtin = builtin
    }
}

/// Mirrors `tag.ts`'s `BaseTagIds` exactly: literal nil-pattern UUIDs
/// `00000000-0000-0000-0000-00000000000N` for N = 1..8, in the plan's listed order (Person,
/// Organization, Company, Event, Place, Area, Project, Task). See that file's doc comment for the
/// full ID-scheme rationale — reproduced here verbatim, not re-derived, since the backend seeds
/// these same literal IDs at workspace-creation time and a client-side mismatch would be silent data
/// corruption, not a caught error.
public enum BaseTagIds {
    public static let person: EntityId = "00000000-0000-0000-0000-000000000001"
    public static let organization: EntityId = "00000000-0000-0000-0000-000000000002"
    public static let company: EntityId = "00000000-0000-0000-0000-000000000003"
    public static let event: EntityId = "00000000-0000-0000-0000-000000000004"
    public static let place: EntityId = "00000000-0000-0000-0000-000000000005"
    public static let area: EntityId = "00000000-0000-0000-0000-000000000006"
    public static let project: EntityId = "00000000-0000-0000-0000-000000000007"
    public static let task: EntityId = "00000000-0000-0000-0000-000000000008"
}

/// Mirrors `tag.ts`'s `BASE_TAGS`: the 8 Base Tags, seeded once at workspace creation as immutable
/// `builtin: true` rows. None have parents — the Base Tags are the roots of the supertag DAG.
public let BASE_TAGS: [Tag] = [
    Tag(id: BaseTagIds.person, name: "Person", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.organization, name: "Organization", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.company, name: "Company", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.event, name: "Event", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.place, name: "Place", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.area, name: "Area", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.project, name: "Project", parentIds: [], builtin: true),
    Tag(id: BaseTagIds.task, name: "Task", parentIds: [], builtin: true)
]
